### PR TITLE
refactor(storage): increase max supported sstable file size

### PR DESCRIPTION
**Pull request by zwang28**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

This PR addresses two issue:
- It increases the supported SSTable file size from u32::MAX bytes to u64::MAX bytes, as well as u64::MAX key_count.
- It rejects overflowed RestartPoint::offset value of u32 type. Because we both support u32::MAX bytes key, u32::MAX bytes value and u32::MAX kilobytes block_size, RestartPoint::offset can overflow.

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: src/storage/src/hummock/sstable/mod.rs [hunk 1]
```diff
@@ -71,7 +72,7 @@ use crate::store::ReadOptions;
 
 const DEFAULT_META_BUFFER_CAPACITY: usize = 4096;
 const MAGIC: u32 = 0x5785ab73;
-const VERSION: u32 = 1;
+const VERSION: u32 = 2;
```
Comment (wenym1) [inline src/storage/src/hummock/sstable/mod.rs:75]: I suggest we maintain a change log on the version here. Each change log item may include the following item:
1. motivation (maybe a issue link, or plain text) 
2. PR number
3. The updated fields in any struct. (type changed, added or removed).

File: src/storage/src/hummock/sstable/mod.rs [hunk 2]
```diff
@@ -411,16 +417,16 @@ impl SstableMeta {
 
     pub fn encode_to(&self, buf: &mut Vec<u8>) {
         let start_offset = buf.len();
-        buf.put_u32_le(self.block_metas.len() as u32);
+        buf.put_u32_le(u32::try_from(self.block_metas.len()).unwrap());
```
Comment (wenym1) [inline src/storage/src/hummock/sstable/mod.rs:420]: May extract the logic of `u32::try_from(..).unwrap()` to a common method named like `assert_u32`

File: src/storage/src/hummock/sstable/mod.rs [hunk 3]
```diff
@@ -442,9 +448,6 @@ impl SstableMeta {
 
         cursor -= 4;
         let version = (&buf[cursor..cursor + 4]).get_u32_le();
```
Comment (wenym1) [inline src/storage/src/hummock/sstable/mod.rs:450]: We may check `version <= VERSION` here?

File: src/storage/src/hummock/sstable/mod.rs (hunk 1 above)
Comment (hzxa21) [reply src/storage/src/hummock/sstable/mod.rs:75]: > I suggest we maintain a change log on the version here. Each change log item may include the following item:
> 1. motivation (maybe a issue link, or plain text) 
> 2. PR number
> 3. The updated fields in any struct. (type changed, added or removed).

+1

File: src/storage/src/hummock/sstable/block.rs [hunk 4]
```diff
@@ -485,7 +485,7 @@ impl BlockBuilder {
         };
 
         let diff_key = if self.entry_count % self.restart_count == 0 || type_mismatch {
-            let offset = self.buf.len() as u32;
+            let offset = u32::try_from(self.buf.len()).unwrap();
```
Comment (MrCroxx) [inline src/storage/src/hummock/sstable/block.rs:488]: Can block size exceeds `u32`? If yes, `offset` here is supposed to be `u64`, too.

**Review by MrCroxx — COMMENTED** [review]
General LGTM. With this PR, when should we `u32` or `u64` in `Block` or `Sstable`? Can `Block` also exceed `u32`?

**Comment by zwang28** [comment]
After some discussion, we decide to panic early on overflow, instead of this PR.

**Pull request closed without merging**